projects/28_gltf/main.cpp

```diff
@@ -0,0 +1,837 @@
+    }
+
+    // This is to simplify the pipeline creation for now. Need to revisit later.
+    PPX_ASSERT_MSG(material.has_pbr_metallic_roughness, "Only PBR metallic roughness supported for now.");
+
+    grfx::PipelineInterfaceCreateInfo piCreateInfo = {};
+    piCreateInfo.setCount                          = 1;
+    piCreateInfo.sets[0].set                       = 0;
+    piCreateInfo.sets[0].pLayout                   = mSetLayout;
+    PPX_CHECKED_CALL(pDevice->CreatePipelineInterface(&piCreateInfo, &pOutput->pInterface));
+
+    grfx::GraphicsPipelineCreateInfo2 gpCreateInfo = {};
+    gpCreateInfo.VS                                = {mVertexShader.Get(), "vsmain"};
+    gpCreateInfo.PS                                = {mPixelShader.Get(), "psmain"};
+    // FIXME: assuming all primitives provides POSITION, UV, NORMAL and TANGENT. Might not be the case.
+    gpCreateInfo.vertexInputState.bindingCount      = 4;
+    gpCreateInfo.vertexInputState.bindings[0]       = mPrimitives[0].mesh->GetDerivedVertexBindings()[0];
+    gpCreateInfo.vertexInputState.bindings[1]       = mPrimitives[0].mesh->GetDerivedVertexBindings()[1];
+    gpCreateInfo.vertexInputState.bindings[2]       = mPrimitives[0].mesh->GetDerivedVertexBindings()[2];
+    gpCreateInfo.vertexInputState.bindings[3]       = mPrimitives[0].mesh->GetDerivedVertexBindings()[3];
+    gpCreateInfo.topology                           = grfx::PRIMITIVE_TOPOLOGY_TRIANGLE_LIST;
+    gpCreateInfo.polygonMode                        = grfx::POLYGON_MODE_FILL;
+    gpCreateInfo.cullMode                           = grfx::CULL_MODE_BACK;
+    gpCreateInfo.frontFace                          = grfx::FRONT_FACE_CCW;
+    gpCreateInfo.depthReadEnable                    = true;
+    gpCreateInfo.depthWriteEnable                   = true;
+    gpCreateInfo.blendModes[0]                      = grfx::BLEND_MODE_NONE;
+    gpCreateInfo.outputState.renderTargetCount      = 1;
+    gpCreateInfo.outputState.renderTargetFormats[0] = pSwapchain->GetColorFormat();
+    gpCreateInfo.outputState.depthStencilFormat     = pSwapchain->GetDepthFormat();
+    gpCreateInfo.pPipelineInterface                 = pOutput->pInterface;
+
+    PPX_CHECKED_CALL(pDevice->CreateGraphicsPipeline(&gpCreateInfo, &pOutput->pPipeline));
+
+    pOutput->textures.resize(3);
+    if (material.pbr_metallic_roughness.base_color_texture.texture == nullptr) {
+        float3 color = glm::make_vec3(material.pbr_metallic_roughness.base_color_factor);
+        LoadTexture(ColorToBitmap(color), pQueue, &pOutput->textures[0]);
+    }
+    else {
+        const auto& texture_path = material.pbr_metallic_roughness.base_color_texture;
+        LoadTexture(gltfFolder, texture_path, pQueue, pTextureCache, &pOutput->textures[0]);
+    }
+
+    if (material.normal_texture.texture == nullptr) {
+        LoadTexture(ColorToBitmap(float3(0.f, 1.f, 0.f)), pQueue, &pOutput->textures[1]);
```

Comment:
Was supposed to be `(0, 0, 1)` fixed! The goal was to have a neutral normal (tangent space) in case we don't have more info than the triangle's normal.
Setting it to `0, 0, 0` would not look.
3 screenshots: with normal map, without (0, 0, 1), and without (0, 0, 0):
![normal](https://user-images.githubusercontent.com/4282252/227889958-c302541f-e54c-4f8a-91d5-9242878e9baf.png)
![no-normal](https://user-images.githubusercontent.com/4282252/227889979-4f2cec9c-3990-4092-be32-b3c4ffda73f5.png)
![all-zero](https://user-images.githubusercontent.com/4282252/227889995-66e08597-32bc-426e-97ec-d89fbab57978.png)